this notebook is to be used for CS379 for david Finger unit 5. created on 10/14/2021.  this program will create a decision tree algorithm and display a visualisation of the predicted results.
the dataset chosen is a bank note autentication dataset. dataset taken from "Decision Trees in Python with Scikit-Learn. (2018, February 28). Stack Abuse. https://stackabuse.com/decision-trees-in-python-with-scikit-learn/"


In [36]:
import numpy as NM
import pandas as PD

import matplotlib.pyplot as PLT #provides output for visualization

from sklearn.model_selection import train_test_split #imports sklearn models for training and testing

from sklearn.tree import DecisionTreeClassifier #imports decision tree for use.

from sklearn.metrics import classification_report, confusion_matrix

the above code segment handles all imports for the jupyter file.

In [37]:
BNA = PD.read_csv('bill_authentication.csv') #loads the dataset into pandas. BNA is Bank Note Authentication

print (BNA.shape) #gives us the shape of the dataset
print() #gives us a space in the code to seperate result visualizations
BNA.head() #prints the head of the dataset so we can see raw data.



(1372, 5)



Variance  Skewness  Curtosis  Entropy  Class
0   3.62160    8.6661   -2.8073 -0.44699      0
1   4.54590    8.1674   -2.4586 -1.46210      0
2   3.86600   -2.6383    1.9242  0.10645      0
3   3.45660    9.5228   -4.0112 -3.59440      0
4   0.32924   -4.4552    4.5718 -0.98880      0

the above code segment loads the dataset into pandas for reading and displays basic information about the dataset.

In [38]:
#splitting and training the dataset begins here


X = BNA.drop('Class', axis = 1) #drops class as it is a unneeded column
Y = BNA['Class'] #puts class as the only column in Y


X_Train, X_Test, Y_Train, Y_Test = train_test_split(X, Y, test_size = .2) #sets up data to be split for training and testing

Tree = DecisionTreeClassifier() #sets the desision tree to its variable we will use.

Tree.fit(X_Train, Y_Train) #puts data into splits

Y_Predict = Tree.predict(X_Test) #sets a prediction for the datset

print (Y_Predict) #prints the prediction

[0 1 0 0 0 0 0 0 1 0 1 0 0 1 0 0 0 1 1 0 1 1 0 1 0 0 1 1 0 0 1 0 1 0 1 1 1
 1 1 0 0 0 0 1 0 0 0 0 1 1 0 1 1 0 0 0 1 1 1 1 0 1 1 0 1 0 0 1 0 1 0 0 1 1
 0 0 0 1 0 0 0 1 0 1 0 0 1 0 0 0 1 0 0 0 0 0 0 1 1 1 0 1 0 0 1 0 1 1 0 0 0
 1 0 1 1 0 0 0 1 0 1 0 0 1 1 0 0 0 0 0 1 1 1 0 1 0 0 0 0 0 1 1 1 0 0 1 0 0
 0 0 1 0 0 0 0 1 1 1 0 0 1 0 1 0 1 0 0 1 1 1 0 0 1 1 0 0 1 0 1 1 0 0 0 1 1
 0 1 0 0 0 0 0 0 0 1 0 0 1 0 0 1 1 1 1 0 1 0 1 0 1 0 0 1 1 0 0 0 1 1 0 0 0
 1 1 0 0 0 0 0 0 1 0 0 1 1 1 1 0 1 0 0 1 1 1 1 1 1 1 1 0 0 1 1 1 1 1 0 1 0
 1 1 0 1 0 0 1 0 0 0 1 0 0 0 1 1]


the above code starts the data training and testing.  splitting the data into 80% training and 20% testing.  then creates a predition for the dataset.

In [39]:
print (confusion_matrix(Y_Test, Y_Predict)) #prints a confusion matrixs from the prediction against recorded data.

print (classification_report (Y_Test, Y_Predict)) #shows the average accuricy of predicted values to recorded data.

[[153   6]
 [  1 115]]
              precision    recall  f1-score   support

           0       0.99      0.96      0.98       159
           1       0.95      0.99      0.97       116

    accuracy                           0.97       275
   macro avg       0.97      0.98      0.97       275
weighted avg       0.98      0.97      0.97       275



these results show that this program has a very high accuricy when it comes to detecting fake notes that come into the system.  in 275 tests that the prediction created it was able to correctly identify it 97% of the time as the correct thing, that being either a true note or a fake note.